api/src/Page/Sample.php

```diff
@@ -1984,24 +1985,45 @@ function _update_sample_group() {
             }
         }
 
-        function _add_sample_to_group() {
-            if (!$this->has_arg('BLSAMPLEID')) $this->_error('No sample specified');
+        function _add_sample_to_group () {
+            $sample = array('BLSAMPLEID' => $this->arg('BLSAMPLEID'), 'BLSAMPLEGROUPID' => $this->arg('BLSAMPLEGROUPID'));
+            $sample_group_result = $this->_save_sample_to_group($sample);
+            $this->_output($sample_group_result);
+        }
+
+        function _bulk_update_samples_in_group () {
+            if (!is_array($this->args)) return $this->_error('Samples for group should be a array');
+            
+            $this->db->start_transaction();
+            $collection = array();
+            foreach ($this->arg('collection') as $sample) {
+                $sample_update_result = $this->_save_sample_to_group($sample);
+
+                array_push($collection, $sample_update_result);
+            }
+
+            $this->db->end_transaction();
+            $this->_output($collection);
+        }
+
+        function _save_sample_to_group($sample_data) {
+            if (!isset($sample_data['BLSAMPLEID'])) $this->_error('No sample specified');
```

Comment:
Rather than an error here it would be better to return null.
Then all the return paths (_error  or _output) are in the calling function which is cleaner.
You would need to check the return value in bulk_update to error if one of the calls to _save_sample_to_group returns null and then _error()

Reply:
@drnasmith since we are also using this for the bulk update which is run as a DB transaction, do we still need to return null or throw an error to rollback the transaction?

Reply:
Yes, return null from _save_sample_to_group. That method will return data or null.
Then in the calling function _bulk_update_samples_in_group, check for the null response.
If null throw the error.
_bulk_update_samples_in_group will either call _output or _error.

I note that _do_add_sample uses the same pattern as you had - i.e. throws errors in the middle of the code. I think its cleaner to use functions that return data or null, then check their result. It keeps all the return paths at the same level. Does that make sense?